### PR TITLE
2016-04-12 adoption & new API design

### PR DESCRIPTION
**Pull request by dreymonde**

### CODE REVIEW

File: Sources/File.swift [hunk 1]
```diff
@@ -114,45 +114,45 @@ public final class File: Stream {
...
             filereadlh(file, $0.baseAddress, lowWaterMark, highWaterMark, deadline.int64milliseconds)
         }
 
-        try FileError.assertNoReceiveErrorWithData(data, bytesProcessed: bytesProcessed)
+        try FileError.assertNoReceiveError(with: data, bytesProcessed: bytesProcessed)
         return Data(data.prefix(bytesProcessed))
     }
 
-    public func read(deadline deadline: Double = .never) throws -> Data {
+    public func read(_ deadline: Double = .never) throws -> Data {
```
Comment (dreymonde) [inline Sources/File.swift]: Underscore here is not needed, I think. `read(5.seconds.fromNow())` looks really strange.

**Comment by dreymonde** [comment]
Also, I think I understand why we’re having this conceptual problems about `File` which can `send` and `receive`. Because we call it `File` when we should not. `File` is oversimplifying. Maybe we should call it `FileHandle`, or even `FileStream`, and `send`/`receive` will make a lot more sense then.

File: Sources/File.swift [hunk 2]
```diff
@@ -188,14 +188,10 @@ extension File {
         return filedetach(file)
     }
 
-    public func close() -> Bool {
-        if closed {
-            return false
-        }
-
+    public func close() throws {
+        guard !closed else { throw ClosableError.alreadyClosed }
```
Comment (scottbyrns) [inline Sources/File.swift:211]: I don't think we would throw exceptions if the connection is already closed. We should reserve it for times when the stream doesn't wish to be closed yet.

Comment (antonmes) [inline Sources/File.swift:211]: you mean if it has some data? i think it should flush data on close, and only rethrow flush errors.

Comment (antonmes) [inline Sources/File.swift:211]: @dreymonde agree, we should move send/receive to FileStream and make File simple with read/write functions

**Comment by dreymonde** [comment]
@antonmes that’s what we were having _before_ this PR :)
The problem is `File` and `FileStream` are basically the same thing.

**Comment by antonmes** [comment]
Ok, then let's discuss `File` and `FileStream` :) Why would we merge them?
`FileStream` is about reading/writing but `File` itself can have `attributes` and other specific api.
I think we should separate them back.

**Comment by scottbyrns** [comment]
A file seems like a type that describes the file where an open file provides a stream.

File has a path, an extension, a size, etc.

A FileStream is an open file that can be read from and written to

**Comment by paulofaria** [comment]
My proposal is to add a new protocol to `C7` called `StreamRepresentable`

``` swift
public protocol StreamRepresentable {
   var stream: Stream { get }
}
```

Then we create an internal class inside `File` module called `FileStream`.
This class would receive a `File` class in its initializer and would translate all calls from `Stream` API to `File` API. `File` class could then expose `FileStream` as a `Stream` using the `stream` property. This way the only public types in `File` module would be `File` class and probably `FileError` enum.

**Comment by paulofaria** [comment]
We can implement this proposal without `StreamRepresentable` being merged. Once it gets merged. We just make `File` adopt it and everything will continue to work.

**Comment by scottbyrns** [comment]
So if we use StreamRepresentable the idea is that a file can be jus a set of describing properties but an extension property getter could exist that uses that information to create a stream?

**Comment by paulofaria** [comment]
I think we can leave all `read` and `write` operations on `File`. But if you think it's best to move all of this functionality to `FileStream`, might be a good idea. Don't know.

**Comment by antonmes** [comment]
File should have `read` and `write` functions

**Pull request merged**